include/robot_interfaces/logger.hpp

```diff
@@ -0,0 +1,287 @@
+///////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <fstream>
+#include <iostream>
+
+#include <chrono>
+
+#include <Eigen/Eigen>
+
+#include <mpi_cpp_tools/basic_tools.hpp>
+#include <mpi_cpp_tools/dynamical_systems.hpp>
+#include <mpi_cpp_tools/math.hpp>
+
+#include <real_time_tools/process_manager.hpp>
+#include <real_time_tools/thread.hpp>
+#include <real_time_tools/threadsafe/threadsafe_timeseries.hpp>
+#include <real_time_tools/timer.hpp>
+
+#include <robot_interfaces/loggable.hpp>
+#include <robot_interfaces/finger_types.hpp>
+
+namespace robot_interfaces
+{
+
+class FingerLogger
+{
+public:
+
+    FingerTypes::Action applied_action;
+    FingerTypes::Action desired_action;
+    FingerTypes::Observation observation;
+    typedef FingerTypes::Status Status;
+    Status status;
+
+    std::shared_ptr<robot_interfaces::RobotData<FingerTypes::Action, FingerTypes::Observation, Status>>
+        logger_data_;
+
+    int block_size_;
+    long int index_;
+    int i, j, loop_iterator_;
+
+    bool stop_was_called_;
+
+    std::ofstream output_file_;
+    std::string output_file_name_;
+    std::vector<std::string> header;
+    std::string temp;
+
+    std::vector<std::string> observation_name_ = observation.get_name();
+    std::vector<std::string> applied_action_name_ = applied_action.get_name();
+    std::vector<std::string> desired_action_name_ = desired_action.get_name();
+
+    std::vector<std::vector<double>> observation_data_ = observation.get_data();
+    std::vector<std::vector<double>> applied_action_data_ = applied_action.get_data();
+    std::vector<std::vector<double>> desired_action_data_ = desired_action.get_data();
+
+    FingerLogger(std::shared_ptr<
+                     robot_interfaces::RobotData<FingerTypes::Action, FingerTypes::Observation, Status>>
+                     robot_data,
+                 int block_size,
+                 std::string filename)
+        : block_size_(block_size),
+          stop_was_called_(false),
+          logger_data_(robot_data),
+          output_file_name_(filename)
+    {
+        thread_ = std::make_shared<real_time_tools::RealTimeThread>();
+    }
+
+    virtual ~FingerLogger()
+    {
+        stop_was_called_ = true;
+        stop();
+        thread_->join();
+
+    }
+
+    void get_header()
+    {
+      //using the newest index inofrmation in logger_data to extract information on the fields to be logged
+      applied_action = (*logger_data_->applied_action)[logger_data_->applied_action->newest_timeindex()];
+      desired_action = (*logger_data_->desired_action)[logger_data_->desired_action->newest_timeindex()];
+      observation = (*logger_data_->observation)[logger_data_->observation->newest_timeindex()];
+      status = (*logger_data_->status)[logger_data_->status->newest_timeindex()];
+
+      header.push_back("#");
+      header.push_back("[Timestamp]");
+      header.push_back("(Current Index)");
```

Comment:
I just did that because it looked nice to me, seemed like differentiating the time information in the header looked cleaner. No specific reason. Will remove them if I should.